### ISSUE TITLE
Handling of table-blocks in dictionary is wrong.

### ISSUE THREAD

**Issue by Markismus** [koreader/koreader#9626]
I've been happily hacking away at converting dictionaries and optimizing them for Koreader and [someone stumbled on a problem](https://github.com/Markismus/PocketBookDic/issues/4): After a `<table>`-block the text entry gets cut off. That 's not the way to handle HTML that you can't parse, right?

This is part of the original PDF:
![image](https://user-images.githubusercontent.com/5269101/195340804-e1e1019b-fd6c-43f1-8e96-b857ffe3dace.png)
This is the result in Goldendict:
![image](https://user-images.githubusercontent.com/5269101/195341233-d6f84c89-9366-4260-9c0e-342cee8b098b.png)
This is the result in Koreader:
![image](https://user-images.githubusercontent.com/5269101/195340907-8b04780a-25a5-4c60-9702-eb2322456a6d.png)
And this is the prettyfied HTML p-block:
```html
<p><span class="font29">Les conclusions sont présentées dans une série de tableaux dont deux sont reproduits ci-après (avec quelques modifications purement formelles). Le tableau I (pris p. 133) indique les positions possibles (-3, -2, -1, +1, +2, +3, +4), par rapport au sujet (S) et au verbe (V), des termes ayant la fonction de complément d’objet direct (COD) ou indirect (COI), d’attribut (A), de complément d’agent (CA), de complément circonstanciel (CC) :</span>
    <p><span class="font1" style="font-weight:bold;font-variant:small-caps;">Tableau</span><span class="font2" style="font-weight:bold;"> I.</span>
        <table border="1">
            <tr>
                <td style="vertical-align:middle;">
                    <p><span class="font2" style="font-weight:bold;">-3</span>
                </td>
                <td style="vertical-align:middle;">
                    <p><span class="font2" style="font-weight:bold;">— 2</span>
                </td>
                <td style="vertical-align:middle;">
                    <p><span class="font2" style="font-weight:bold;">-1</span>
                </td>
                <td style="vertical-align:middle;">
                    <p><span class="font29">s</span>
                </td>
                <td style="vertical-align:middle;">
                    <p><span class="font24" style="font-weight:bold;">.....</span>
                </td>
                <td style="vertical-align:middle;">
                    <p><span class="font29">v</span>
                </td>
                <td style="vertical-align:middle;">
                    <p><span class="font26">+ 1</span>
                </td>
                <td style="vertical-align:middle;">
                    <p><span class="font2" style="font-weight:bold;">+2</span>
                </td>
                <td style="vertical-align:middle;">
                    <p><span class="font2" style="font-weight:bold;">4-3</span>
                </td>
                <td style="vertical-align:middle;">
                    <p><span class="font2" style="font-weight:bold;">+4</span>
                </td>
            </tr>
            <tr>
                <td>
                    <p><span class="font29">cc</span>
                </td>
                <td>
                    <p><span class="font2" style="font-weight:bold;">CC</span>
                </td>
                <td>
                    <p><span class="font2" style="font-weight:bold;">CC</span>
                </td>
                <td></td>
                <td>
                    <p><span class="font2" style="font-weight:bold;">CC</span>
                        <p><span class="font2" style="font-weight:bold;text-decoration:underline;">COD COI</span>
                </td>
                <td></td>
                <td style="vertical-align:middle;">
                    <p><span class="font2" style="font-weight:bold;">CC </span><span class="font2" style="font-weight:bold;text-decoration:underline;">COD COI </span><span class="font2" style="font-weight:bold;">A CA</span>
                </td>
                <td>
                    <p><span class="font2" style="font-weight:bold;">CC</span>
                        <p><span class="font2" style="font-weight:bold;">COD</span>
                            <p><span class="font2" style="font-weight:bold;">COI</span>
                                <p><span class="font2" style="font-weight:bold;">A</span>
                </td>
                <td>
                    <p><span class="font2" style="font-weight:bold;">CC</span>
                </td>
                <td>
                    <p><span class="font2" style="font-weight:bold;">CC</span>
                </td>
            </tr>
        </table>
        <p>
```

An easy enough approach is to protect the entries for koreader with a gsub-command in the accompagnying lua-file deleting the tables or changing them to more basic structures or to an ASCIIi rendering. _Does anyone know of a succesful approach for that?_

However, it's not the way the dictionary should handle tables. Either it removes them and displays the rest or it creatively rerenders them.

**Comment by poire-z** [comment]
May be our old MuPDF (used for rendering HTML dicts results) does not handle well `<table>` among inline elements (it is inside the `<P>` with a bunch of `<SPAN>`), which is not an easy thing to handle - so MuPDF may consider/reset every subelements to be inline.
You could try replacing `<table>` with `</p><table>` (or `</p></div></p></div><table>` to be sure to close many :) and see what happens.

**Comment by Markismus** [comment]
No luck. I tried `html = html:gsub("(<table[^>]+>)", "</p></div></p></div>%1")`, restarted Koreader and removed a stop-tag and repeated. No change in display.

Trying `html = html:gsub("<table[^>]+>.-</table>", "")` I got rid of all the tables. Just removing the tags resulted in odd outer vertical lines for the rest of the article.

**Comment by poire-z** [comment]
Going first with `</p></div></p></div>` might not be the best :) I remember MuPDF expects balanced tags (so, it expects XHTML).
Also, a real HTML parser has some rules, like closing any `<P>` when meeting a `<TABLE>`, which I guess MuPDF may not have implemented it all perfectly.

So, try with `</p><table>` first, but also add `</table>` => `</table><p>` (reopen what you closed).
Although in your sample, the first `<P>` is not explicitely closed. And a new `<P>` starts after the table.

Also (and may be first) check if MuPDF is really capable of showing TABLEs :)

Also, don't expect MuPDF (and even more our old version) to be a top HTML renderer. It has bugs (ie. https://github.com/koreader/koreader/blob/8508689cea0bf7490e70b7c221336f4912d73c3b/frontend/ui/widget/htmlboxwidget.lua#L51-L54) that you may have to 1) find out there are  bugs 2) workaround :/

(We may fix/improve crengine, our EPUB renderer - but we try to not touch MuPDF much so we can keep in sync with upstream, although it's been a few years we have not updated - https://github.com/koreader/koreader-base/pull/1203)

Just noticed:
> And this is the prettyfied HTML p-block:

You have `<P>` imbricated - which is not really proper HTML (no P inside P). May be it's just your prettified or you copy and pasting.

**Comment by Frenzie** [comment]
Our MuPDF supports tables, though I'm not sure if it supports all colspan/rowspan intricacies. 1.18 added an HTML parser, so that one *may* be able to handle the spaghetti above by terminating the paragraph before starting the table.

**Comment by bousnah** [comment]
Just to point out,again, that the same files of this dictionary under Koreader are used under goldendict.  Goldendict sees completely Tableau 1 as well as the rest of the article under "ordre".   Something is not working under koreader that works under goldendict even though the files are the same.  The goldendict that I have is under windows 11

**Comment by poire-z** [comment]
May be GoldenDict can show nice dicts results, but give it an EPUB and see ! :)
KOReader is a book reader - we do _what we can_ with dictionaries (and usually good enough as most dicts output text or simple HTML).
On Windows or Android, good Dict tooks like GoldenDict can use the system's or Edge/Chrome subcomponents to display complex HTML nicely.
We have to use MuPDF, with its limitations. And we also don't care about showing images in dict results, which would obviously need more work).

**Comment by Frenzie** [comment]
The above image is also what we'll get with MuPDF 1.18 and higher with the original HTML in the future.

**Comment by Frenzie** [comment]
If you get rid of all the internal paragraphs then you'll see the output you expect.

![Screenshot_20221012_183130](https://user-images.githubusercontent.com/202757/195398630-c4c3de22-b574-47c4-be78-c7ad4db61e2e.png)

**Comment by Markismus** [comment]
When I manually replace the tags, Firefox displays is as such:
![image](https://user-images.githubusercontent.com/5269101/195420063-9ff679cc-6413-4f81-bfcd-bc38ac4634a6.png)
(Without removing the `<p>`-tags it's still horizontal, but more garbled.)

In lua, I wrote it as follows:
```lua
html = html:gsub("(<table[^>]+>.-)</?p[^>]*>(.-</table>)", "%1%2")
html = html:gsub("<table[^>]+>", '<div style="display:table;>')
html = html:gsub("<tr[^>]+>", '<div style="display:table-row;>')
html = html:gsub("<td[^>]+>", '<div style="display:table-cell;>')
html = html:gsub("</table>|</tr>|</td>", '</div>')
```
It's better than deletion of the table, however Koreader still displays all vertically:
![image](https://user-images.githubusercontent.com/5269101/195420671-8cf28eeb-d0ae-4ee6-a903-52f3c2a4f5dd.png)

What am I doing wrong? Or is to be expected of mu-pdf 1.3?

**Comment by poire-z** [comment]
MuPDF probably does not support `display:table-row` - so don't try doing anything fancy with CSS from after year 1998 :)
Just try to get it out of `<P>`. May be try first to remove everything that's before `<TABLE>` and after `</TABLE>` and see if you can get it to at least display a standalone table.

**Comment by Frenzie** [comment]
I meant in;
```html
                <td style="vertical-align:middle;">
                    <p><span class="font2" style="font-weight:bold;">-3</span>
                </td>
```

Change it to:
```html
                <td style="vertical-align:middle;">
                    <span class="font2" style="font-weight:bold;">-3</span>
                </td>
```

(Besides the obvious of properly closing tags.)

**Comment by Markismus** [comment]
I know, but the only appear in tables. So `html = html:gsub("(<table[^>]+>.-)</?p[^>]*>(.-</table>)", "%1%2")` should be good, right? Or should I wrap it in an extra `while`-loop in case the search pattern isn't reset to the start of the string after a substitution:
```lua
count = 1
while count do
html, count  = html:gsub("(<table[^>]+>.-)</?p[^>]*>(.-</table>)", "%1%2")
end
```

**Comment by Markismus** [comment]
Also, how do I reduce image-size in posts using markdown?

**Comment by Frenzie** [comment]
> How do I dump the string html to a file within Koreader?

io.write? ^_^

> how do I reduce image-size in posts using markdown?

By not using Markdown. :-P

`<img src="leave what came from the upload" width=45%>`

**Comment by Frenzie** [comment]
> How do I dump the string html to a file within Koreader?

But hold on, what are you trying to do? It seems to me it's orders of magnitude simpler to do something like:

```lua
table_stuff = [[ -- or io.open, same difference
<table etc.
]]

output = table_stuff:gsub…
print(output)
```

Considering you're not really doing anything with the program at all.

**Comment by bousnah** [comment]
I raised the issue of the table within the linguistic article contained within a headword definition.  I want to point out that tables will not occur in the definitions proper but only,if there is to be a table/s displayed,after the "grammaire et linguistique" line appears.  I think it would be nice to have those tables but more importantly is the text, so if you can retain the text before and after a table I think that would certainly be satisfactory.  One can always refer to goldendict to see the tableau(table).   The main concern of course is having the complete text of the definitions that, as I said, do not contain tables. The grammaire and linguistique, it should be noted, always come at the end of a defintion-no exceptions. I have also raised the important issue of finding headwords that sometimes have "1." before the headword.  Can you correct the search engine to find any headwords that have a "1." before the headword?  I gave examples in a comment earlier to Markismus of "1. avoir" and "1. porter"  .  "Avoir" is one of the headwords that commands one of the longerst definition in this dictionnary and it goes missing because "1." is prefixed.  A look at the  pdf shows how the word and definiton is listed.  A lot of words will go missing if "1."  is not found.  I addressed this in another comment.

**Comment by Markismus** [comment]
![image](https://user-images.githubusercontent.com/5269101/197227821-387eff67-da5f-42fe-b6e0-ad45d4502f41.png)

After I created a nice reconstruction of tables with div-blocks, vertical bars and horizontal lines, I realized, that I hadn't tested @poire-z 's suggestion of just getting rid of the p-blocks....and that also worked. :) Just three lines.
```lua
-- Remove p start- and stoptags from table
while html:find("(<table[^>]+>.-)</?p[^>]*>(.-</table>)") do
  html  = html:gsub("(<table[^>]+>.-)</?p[^>]*>(.-</table>)", "%1%2")
end
-- If malformed tables continue to stop Koreader from displaying beyond the start 
-- of a table, please uncomment the following lines to convert them to a
-- simulation of a table constructed from div-blocks, vertical bars and 
-- horizontal lines.

-- html = html:gsub("<table[^>]+>", '<p></p><div style="display:block;">')
-- html = html:gsub("</table>", '</div><p></p>')
-- html = html:gsub("<tr[^>]*>", '<div style="display:block;">')
-- html = html:gsub("<td[^>]*>", '<div style="display:inline;">|') -- ｜
-- html = html:gsub("</td>", '</div>')
-- html = html:gsub("</tr>", '|</div><hr style="height:1px;color:black;" />')
```

**Comment by bousnah** [comment]
Would you kindly tell me if, in fact, you have succeeded in
re-constructing the tableaux/tables.   It appears that this issue has been
closed yet,  I have the impression that maybe this problem was solved.
Would there be another revision of the dictionnary if the tables can indeed
appear in the text? The current revision is 20-10-2022.

cordially

**Comment by poire-z** [comment]
To answers @bousnah 2 points:

> I raised the issue of the table within the linguistic article contained within a headword definition. I want to point out that tables will not occur in the definitions proper but only,if there is to be a table/s displayed,after the "grammaire et linguistique" line appears. [...]

The tool we use to render HTML definitions (MuPDF) has some limitations, and won't display well some HTML.
This is not something we (KOReader) want to fight.
But we allow users to tweak the HTML we get from the dictionary before giving it to our renderer, with a Lua script.
And that is exactly what @Markismus is doing - so you will probably benefit from his work (if it's the same dictionary you're looking at).
Fixing such dictionary output (too complex or bad HTML) is not our (KOReader) job - but we provide hooks for users to help themselves.

> I have also raised the important issue of finding headwords that sometimes have "1." before the headword. Can you correct the search engine to find any headwords that have a "1." before the headword? I gave examples in a comment earlier to Markismus of "1. avoir" and "1. porter" . "Avoir" is one of the headwords that commands one of the longerst definition in this dictionnary and it goes missing because "1." is prefixed. A look at the pdf shows how the word and definiton is listed. A lot of words will go missing if "1." is not found. I addressed this in another comment.

But for this, it is totally not to be fixed by us :) It's your dictionary that is bad. If some definitions starts with "1. avoir" and "2. avoir", they should both be indexed for "avoir": their should only be "avoir" in the index that points to these 2 definitions. If not, it's for you (or the publisher of the dict) to fix the dictionary index. It's not KOReader (or sdvc, the tool we use for the stardict lookup") to reconstruct an index when it is bad - the code can not guess the index is bad, but you the user can.

**Comment by bousnah** [comment]
Thank you for your message.   I think that M Markismus resolved the issue
of the headwords pre-fixed by numbers.
As to tables, from the discussion that I just received today, I had the
impression that maybe tables could indeed be integrated in the
dictionnary.
Perhaps this is not the case and as I commented earlier on this topic it is
more important to include all the text of the linguistic articles than to
have the tableaux/tables if one of these must be sacrificed.
I understand that koreader has limits; I am not able to tweak any code as I
completely lack the know-how to do this.  I am thankful for the others such
as M Markismus who has kindly devoted his time and expertise on this
project.  And I am certainly thankful for those who have made Koreader such
a wonderful tool that has enhance immensely the e-reader experience.

cordially
